File: go/internal/pagou/http.go
package pagou

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"math"
	"net/http"
	"net/url"
	"strconv"
	"strings"
	"time"
)

var (
	retryableStatus  = map[int]struct{}{429: {}, 500: {}, 502: {}, 503: {}, 504: {}}
	idempotentMethod = map[string]struct{}{http.MethodGet: {}, http.MethodHead: {}}
)

// RequestParams describes a single call to the API.
type RequestParams struct {
	Method string
	Path   string
	Query  map[string]any
	Body   any
	// IdempotencyKey is sent as Idempotency-Key and also makes a write retryable
	// on transient failures.
	IdempotencyKey string
	// RequestID is the correlation id echoed as X-Request-Id. Generated when empty.
	RequestID string
	TimeoutMs int
}

// Result carries a decoded payload plus response metadata.
type Result[T any] struct {
	Data      T
	Status    int
	RequestID string
}

// dataEnvelope is the { success, requestId, data } wrapper every non-list
// endpoint returns.
type dataEnvelope[T any] struct {
	Success   bool   `json:"success"`
	RequestID string `json:"requestId"`
	Data      T      `json:"data"`
}

// CursorPage is the envelope for list endpoints. The cursor fields are
// snake_case on the wire.
type CursorPage[T any] struct {
	Success    bool    `json:"success"`
	RequestID  string  `json:"requestId"`
	Data       []T     `json:"data"`
	NextCursor *string `json:"next_cursor"`
	PrevCursor *string `json:"prev_cursor"`
	Total      int     `json:"total"`
}

// Client is a minimal, dependency-light reference client for the Pagou API v2
// built on net/http. It demonstrates the fundamentals every language example
// must show: server-side auth, correlation ids, idempotency keys, timeouts,
// bounded retries for transient failures on idempotent operations, typed errors
// and redacted logging.
type Client struct {
	config Config
	http   *http.Client
}

// New builds a client from the given config.
func New(cfg Config) *Client {
	return NewWithHTTPClient(cfg, &http.Client{})
}

// NewWithHTTPClient builds a client with a caller-supplied *http.Client. Tests
// inject a client whose Transport returns canned responses.
func NewWithHTTPClient(cfg Config, hc *http.Client) *Client {
	return &Client{config: cfg, http: hc}
}

func (c *Client) authHeader() (string, string) {
	// The API key is a server-side secret; it is never read in browser code.
	return "Authorization", "Bearer " + c.config.APIToken
}

func (c *Client) buildURL(path string, query map[string]any) (string, error) {
	base, err := url.Parse(strings.TrimRight(c.config.BaseURL, "/") + "/")
	if err != nil {
		return "", err
	}
	ref, err := url.Parse(strings.TrimPrefix(path, "/"))
	if err != nil {
		return "", err
	}
	u := base.ResolveReference(ref)
	if len(query) > 0 {
		q := u.Query()
		for key, value := range query {
			if s, ok := queryValue(value); ok {
				q.Set(key, s)
			}
		}
		u.RawQuery = q.Encode()
	}
	return u.String(), nil
}

func queryValue(value any) (string, bool) {
	switch v := value.(type) {
	case nil:
		return "", false
	case string:
		return v, true
	case []string:
		if len(v) == 0 {
			return "", false
		}
		return strings.Join(v, ","), true
	case int:
		return strconv.Itoa(v), true
	case bool:
		return strconv.FormatBool(v), true
	default:
		return fmt.Sprintf("%v", v), true
	}
}

func (c *Client) canRetry(method, idempotencyKey string) bool {
	if _, ok := idempotentMethod[method]; ok {
		return true
	}
	// Writes are retried only when an idempotency key guards against duplicates.
	return (method == http.MethodPost || method == http.MethodPut) && idempotencyKey != ""
}

// rawResult holds a decoded-but-unwrapped response body.
type rawResult struct {
	body      json.RawMessage
	status    int
	requestID string
}

func (c *Client) request(p RequestParams) (*rawResult, error) {
	requestID := p.RequestID
	if requestID == "" {
		requestID = newUUID()
	}
	fullURL, err := c.buildURL(p.Path, p.Query)
	if err != nil {
		return nil, networkError("Invalid request URL", requestID, err)
	}
	retryable := c.canRetry(p.Method, p.IdempotencyKey)
	maxAttempts := 1
	if retryable {
		maxAttempts = c.config.MaxRetries + 1
	}

	Log().Info(fmt.Sprintf("→ %s %s", p.Method, fullURL), map[string]any{
		"requestId": requestID,
		"body":      p.Body,
	})

	timeout := time.Duration(p.TimeoutMs) * time.Millisecond
	if timeout <= 0 {
		timeout = time.Duration(c.config.TimeoutMs) * time.Millisecond
	}

	var lastErr error
	for attempt := 0; attempt < maxAttempts; attempt++ {
		result, retry, err := c.attempt(p, fullURL, requestID, retryable, attempt, maxAttempts, timeout)
		if err != nil {
			lastErr = err
			if retry {
				continue
			}
			return nil, err
		}
		return result, nil
	}
	return nil, networkError("Request failed after retries", requestID, lastErr)
}

func (c *Client) attempt(p RequestParams, fullURL, requestID string, retryable bool, attempt, maxAttempts int, timeout time.Duration) (*rawResult, bool, error) {
	ctx, cancel := context.WithTimeout(context.Background(), timeout)
	defer cancel()

	var bodyReader io.Reader
	if p.Body != nil {
		encoded, err := json.Marshal(p.Body)
		if err != nil {
			return nil, false, networkError("Failed to encode request body", requestID, err)
		}
		bodyReader = bytes.NewReader(encoded)
	}

	req, err := http.NewRequestWithContext(ctx, p.Method, fullURL, bodyReader)
	if err != nil {
		return nil, false, networkError("Failed to build request", requestID, err)
	}
	req.Header.Set("Accept", "application/json")
	req.Header.Set("X-Request-Id", requestID)
	req.Header.Set(c.authHeader())
	if p.IdempotencyKey != "" {
		req.Header.Set("Idempotency-Key", p.IdempotencyKey)
	}
	if p.Body != nil {
		req.Header.Set("Content-Type", "application/json")
	}

	resp, err := c.http.Do(req)
	if err != nil {
		isTimeout := errors.Is(err, context.DeadlineExceeded) || ctx.Err() == context.DeadlineExceeded
		if retryable && attempt < maxAttempts-1 {
			sleep(backoff(attempt, ""))
			return nil, true, err
		}
		message := "Network request failed"
		if isTimeout {
			message = "Request timed out"
		}
		return nil, false, networkError(message, requestID, err)
	}
	defer resp.Body.Close()

	responseID := resp.Header.Get("X-Request-Id")
	if responseID == "" {
		responseID = requestID
	}
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		if retryable && attempt < maxAttempts-1 {
			sleep(backoff(attempt, ""))
			return nil, true, err
		}
		return nil, false, networkError("Failed to read response body", requestID, err)
	}

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		if _, ok := retryableStatus[resp.StatusCode]; retryable && ok && attempt < maxAttempts-1 {
			sleep(backoff(attempt, resp.Header.Get("Retry-After")))
			return nil, true, fmt.Errorf("retryable status %d", resp.StatusCode)
		}
		apiErr := toAPIError(resp.StatusCode, decodeAny(body), responseID)
		Log().Warn(fmt.Sprintf("← %d %s %s", resp.StatusCode, p.Method, p.Path), map[string]any{
			"requestId": responseID,
			"code":      apiErr.Code,
		})
		return nil, false, apiErr
	}

	Log().Info(fmt.Sprintf("← %d %s %s", resp.StatusCode, p.Method, p.Path), map[string]any{"requestId": responseID})
	return &rawResult{body: json.RawMessage(body), status: resp.StatusCode, requestID: responseID}, false, nil
}

// RequestData unwraps a { success, requestId, data } envelope to its data.
func RequestData[T any](c *Client, p RequestParams) (Result[T], error) {
	raw, err := c.request(p)
	if err != nil {
		return Result[T]{}, err
	}
	var env dataEnvelope[T]
	if err := json.Unmarshal(raw.body, &env); err != nil {
		return Result[T]{}, networkError("Failed to decode response envelope", raw.requestID, err)
	}
	requestID := env.RequestID
	if requestID == "" {
		requestID = raw.requestID
	}
	return Result[T]{Data: env.Data, Status: raw.status, RequestID: requestID}, nil
}

// RequestCursorPage returns a full cursor page (keeps next/prev cursor and total).
func RequestCursorPage[T any](c *Client, p RequestParams) (Result[CursorPage[T]], error) {
	raw, err := c.request(p)
	if err != nil {
		return Result[CursorPage[T]]{}, err
	}
	var page CursorPage[T]
	if err := json.Unmarshal(raw.body, &page); err != nil {
		return Result[CursorPage[T]]{}, networkError("Failed to decode cursor page", raw.requestID, err)
	}
	return Result[CursorPage[T]]{Data: page, Status: raw.status, RequestID: raw.requestID}, nil
}

func decodeAny(body []byte) any {
	if len(bytes.TrimSpace(body)) == 0 {
		return nil
	}
	var out any
	if err := json.Unmarshal(body, &out); err != nil {
		return string(body)
	}
	return out
}

func backoff(attempt int, retryAfter string) time.Duration {
	if retryAfter != "" {
		if seconds, err := strconv.ParseFloat(retryAfter, 64); err == nil {
			return time.Duration(math.Min(seconds*1000, 5000)) * time.Millisecond
		}
	}
	base := 200 * math.Pow(2, float64(attempt))
	jitter := math.Floor(deterministicJitter(attempt) * 200)
	return time.Duration(math.Min(base+jitter, 5000)) * time.Millisecond
}

// deterministicJitter keeps the reference reproducible without a random source.
func deterministicJitter(attempt int) float64 {
	x := math.Sin(float64(attempt+1)) * 10_000
	return x - math.Floor(x)
}

func sleep(d time.Duration) { time.Sleep(d) }

File: python/pagou/http.py
import json
import math
import time
import uuid
from dataclasses import dataclass
from typing import Any, Generic, Optional, TypeVar
from urllib.parse import urlencode

import httpx

from .config import PagouConfig, load_config
from .errors import NetworkError, to_api_error
from .logger import logger

T = TypeVar("T")

_RETRYABLE_STATUS = {429, 500, 502, 503, 504}
_IDEMPOTENT_METHODS = {"GET", "HEAD"}


@dataclass
class Result(Generic[T]):
    data: T
    status: int
    request_id: Optional[str] = None


@dataclass
class CursorPage(Generic[T]):
    """Cursor envelope for list endpoints.

    ``next_cursor`` / ``prev_cursor`` are snake_case, matching the wire format.
    """

    success: bool
    request_id: str
    data: list
    next_cursor: Optional[str]
    prev_cursor: Optional[str]
    total: int

    @classmethod
    def from_envelope(cls, body: dict) -> "CursorPage":
        return cls(
            success=bool(body.get("success", True)),
            request_id=body.get("requestId", ""),
            data=body.get("data", []) or [],
            next_cursor=body.get("next_cursor"),
            prev_cursor=body.get("prev_cursor"),
            total=int(body.get("total", 0)),
        )


class PagouHttpClient:
    """Minimal, dependency-light reference client for the Pagou API v2.

    It demonstrates the fundamentals every language example must show: server-side
    auth, correlation ids, idempotency keys, timeouts, bounded retries for
    transient failures on idempotent operations, typed errors and redacted logging.
    """

    def __init__(
        self,
        config: Optional[PagouConfig] = None,
        transport: Optional[httpx.BaseTransport] = None,
    ) -> None:
        self.config = config or load_config()
        # A custom transport lets tests inject responses without a network call.
        self._client = httpx.Client(transport=transport)

    def close(self) -> None:
        self._client.close()

    def __enter__(self) -> "PagouHttpClient":
        return self

    def __exit__(self, *_exc: Any) -> None:
        self.close()

    def _auth_header(self) -> tuple[str, str]:
        # The API key is a server-side secret; it is never read in browser code.
        return ("Authorization", f"Bearer {self.config.api_token}")

    def _build_url(self, path: str, query: Optional[dict]) -> str:
        base = self.config.base_url.rstrip("/")
        url = f"{base}/{path.lstrip('/')}"
        if not query:
            return url
        pairs = []
        for key, value in query.items():
            if value is None:
                continue
            if isinstance(value, (list, tuple)):
                if not value:
                    continue
                pairs.append((key, ",".join(str(v) for v in value)))
            else:
                pairs.append((key, str(value)))
        if not pairs:
            return url
        return f"{url}?{urlencode(pairs)}"

    def _can_retry(self, method: str, idempotency_key: Optional[str]) -> bool:
        if method in _IDEMPOTENT_METHODS:
            return True
        # Writes are retried only when an idempotency key guards against duplicates.
        return method in ("POST", "PUT") and bool(idempotency_key)

    def request(
        self,
        method: str,
        path: str,
        *,
        query: Optional[dict] = None,
        body: Any = None,
        idempotency_key: Optional[str] = None,
        request_id: Optional[str] = None,
        timeout_ms: Optional[int] = None,
    ) -> Result:
        request_id = request_id or str(uuid.uuid4())
        url = self._build_url(path, query)
        retryable = self._can_retry(method, idempotency_key)
        max_attempts = (self.config.max_retries + 1) if retryable else 1
        timeout_s = (timeout_ms if timeout_ms is not None else self.config.timeout_ms) / 1000

        logger.info(f"→ {method} {path}", {"requestId": request_id, "body": body})

        last_error: Optional[BaseException] = None
        for attempt in range(max_attempts):
            headers = {"Accept": "application/json", "X-Request-Id": request_id}
            headers[self._auth_header()[0]] = self._auth_header()[1]
            if idempotency_key:
                headers["Idempotency-Key"] = idempotency_key
            content = None
            if body is not None:
                headers["Content-Type"] = "application/json"
                content = json.dumps(body)

            try:
                response = self._client.request(
                    method, url, headers=headers, content=content, timeout=timeout_s
                )
            except httpx.TimeoutException as error:
                last_error = error
                if retryable and attempt < max_attempts - 1:
                    time.sleep(_backoff_seconds(attempt, None))
                    continue
                raise NetworkError("Request timed out", request_id=request_id, cause=error)
            except httpx.HTTPError as error:
                last_error = error
                if retryable and attempt < max_attempts - 1:
                    time.sleep(_backoff_seconds(attempt, None))
                    continue
                raise NetworkError("Network request failed", request_id=request_id, cause=error)

            response_id = response.headers.get("x-request-id") or request_id
            payload = _parse_body(response)

            if response.status_code >= 400:
                if (
                    retryable
                    and response.status_code in _RETRYABLE_STATUS
                    and attempt < max_attempts - 1
                ):
                    time.sleep(_backoff_seconds(attempt, response.headers.get("Retry-After")))
                    continue
                error = to_api_error(response.status_code, payload, response_id)
                logger.warn(
                    f"← {response.status_code} {method} {path}",
                    {"requestId": response_id, "code": error.code},
                )
                raise error

            logger.info(f"← {response.status_code} {method} {path}", {"requestId": response_id})
            return Result(data=payload, status=response.status_code, request_id=response_id)

        raise NetworkError("Request failed after retries", request_id=request_id, cause=last_error)

    def request_data(self, method: str, path: str, **kwargs: Any) -> Result:
        """Unwraps a ``{ success, requestId, data }`` envelope to its ``data``."""
        result = self.request(method, path, **kwargs)
        envelope = result.data if isinstance(result.data, dict) else {}
        return Result(
            data=envelope.get("data"),
            status=result.status,
            request_id=envelope.get("requestId") or result.request_id,
        )

    def request_cursor_page(self, method: str, path: str, **kwargs: Any) -> Result:
        """Returns a full cursor page (keeps ``next_cursor`` / ``prev_cursor`` / ``total``)."""
        result = self.request(method, path, **kwargs)
        body = result.data if isinstance(result.data, dict) else {}
        return Result(
            data=CursorPage.from_envelope(body),
            status=result.status,
            request_id=result.request_id,
        )


def _parse_body(response: httpx.Response) -> Any:
    content_type = response.headers.get("content-type", "")
    text = response.text
    if not text:
        return None
    if "json" in content_type:
        try:
            return json.loads(text)
        except ValueError:
            return text
    return text


def _backoff_seconds(attempt: int, retry_after: Optional[str]) -> float:
    if retry_after:
        try:
            seconds = float(retry_after)
            return min(seconds, 5.0)
        except ValueError:
            pass
    base = 200 * (2**attempt)
    jitter = math.floor(_deterministic_jitter(attempt) * 200)
    return min(base + jitter, 5000) / 1000


def _deterministic_jitter(attempt: int) -> float:
    # Small deterministic jitter keeps the reference reproducible without random.
    x = math.sin(attempt + 1) * 10_000
    return x - math.floor(x)
